// redux-subspace-saga/es/index.js
import _regeneratorRuntime from 'babel-runtime/regenerator';
import { setContext, getContext, takeEvery } from 'redux-saga/effects';
import createSagaMiddleware, { runSaga } from 'redux-saga';
import { subspace, applyToRoot } from 'redux-subspace';

/**
 * Copyright 2017, IOOF Holdings Limited.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

var provideStore = function provideStore(store, options) {
  return function (saga) {
    return (/*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        var _args = arguments;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return setContext({ store: store, sagaMiddlewareOptions: options });

              case 2:
                return _context.delegateYield(saga.apply(undefined, _args), 't0', 3);

              case 3:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      })
    );
  };
};

/**
 * Copyright 2017, IOOF Holdings Limited.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

var createSagaMiddleware$1 = (function (options) {
    var sagaMiddleware = createSagaMiddleware(options);

    var sagaSubspaceMiddleware = function sagaSubspaceMiddleware(store) {
        sagaSubspaceMiddleware.run = function (saga) {
            for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
                args[_key - 1] = arguments[_key];
            }

            return sagaMiddleware.run.apply(sagaMiddleware, [provideStore(store, options)(saga)].concat(args));
        };
        return sagaMiddleware(store);
    };

    return sagaSubspaceMiddleware;
});

var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];

    for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }

  return target;
};

/**
 * Copyright 2017, IOOF Holdings Limited.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

var emitter = function emitter() {
  var subscribers = [];

  function subscribe(sub) {
    subscribers.push(sub);
    return function () {
      subscribers.splice(subscribers.indexOf(sub), 1);
    };
  }

  function emit(item) {
    var arr = subscribers.slice();
    for (var i = 0, len = arr.length; i < len; i++) {
      arr[i](item);
    }
  }

  return {
    subscribe: subscribe,
    emit: emit
  };
};

var subspaced = function subspaced(mapState, namespace) {
  var subspaceDecorator = subspace(mapState, namespace);

  return function (saga) {
    return (/*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        var parentStore, sagaMiddlewareOptions, sagaEmitter, store;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return getContext('store');

              case 2:
                parentStore = _context2.sent;
                _context2.next = 5;
                return getContext('sagaMiddlewareOptions');

              case 5:
                sagaMiddlewareOptions = _context2.sent;
                sagaEmitter = emitter();
                store = _extends({}, sagaMiddlewareOptions, subspaceDecorator(parentStore), {
                  subscribe: sagaEmitter.subscribe
                });


                runSaga.apply(undefined, [store, provideStore(store, sagaMiddlewareOptions)(saga)].concat(args));

                _context2.next = 11;
                return takeEvery('*', /*#__PURE__*/_regeneratorRuntime.mark(function _callee(action) {
                  return _regeneratorRuntime.wrap(function _callee$(_context) {
                    while (1) {
                      switch (_context.prev = _context.next) {
                        case 0:
                          store.processAction(action, sagaEmitter.emit);
                          _context.next = 3;
                          return;

                        case 3:
                        case 'end':
                          return _context.stop();
                      }
                    }
                  }, _callee, this);
                }));

              case 11:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      })
    );
  };
};

/**
 * Copyright 2017, IOOF Holdings Limited.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

var index = (function (options) {
  return applyToRoot(createSagaMiddleware$1(options));
});

export default index;
export { provideStore, subspaced };
